autosklearn/automl.py

```diff
@@ -1834,15 +1834,52 @@ def get_models_with_weights(self):
         return self.ensemble_.get_models_with_weights(self.models_)
 
     def show_models(self):
-        models_with_weights = self.get_models_with_weights()
-
-        with io.StringIO() as sio:
-            sio.write("[")
-            for weight, model in models_with_weights:
-                sio.write("(%f, %s),\n" % (weight, model))
-            sio.write("]")
-
-            return sio.getvalue()
+        """ Returns a dictionary containing models and their information
+            where model_id/run_id is the key """
+
+        ensemble_dict={} 
+        
+        def has_key(rv, key):
+            return rv.additional_info and key in rv.additional_info
+        
+        table_dict={}
+        for rkey, rval in self.runhistory_.data.items():
+            if has_key(rval, 'num_run'):
+                table_dict[rval.additional_info['num_run']]={
+                        'model_id': rval.additional_info['num_run'],
+                        'cost': rval.cost}
+                
+        for i, weight in enumerate(self.ensemble_.weights_):
+            (_, model_id, _) = self.ensemble_.identifiers_[i]
+            table_dict[model_id]['ensemble_weight']=weight
+        
+        table=pd.DataFrame.from_dict(table_dict, orient='index')
+        table.sort_values(by='cost', inplace=True)
+        table['rank']=range(1,len(table)+1)
+
+        for (_, model_id, _), model in self.models_.items():
+            model_dict={} #Empty model dictionary
+                
+            #Inserting rank and ensemble weight
+            model_dict['model_id']=table.loc[model_id]['model_id']
+            model_dict['rank']=table.loc[model_id]['rank']
+            model_dict['ensemble_weight']=table.loc[model_id]['ensemble_weight']
+            
+            # The steps in the models pipeline will be saved in the dictionary as follows: 
+            # 'data_preprocessing': DataPreprocessor,
+            # 'balancing': Balancing,
+            # 'feature_preprocessor': FeaturePreprocessorChoice,
+            # 'classifier': ClassifierChoice -> autosklearn wrapped model
+            for step in model.steps:
+                model_dict[step[0]]=step[1]
+            
+            #Adding sklearn model to the model dictionary
+            model_dict['sklearn_model']=model.steps[-1][1].choice.estimator
```

Comment:
This is sort of duplicated from including the step above but I agree with your choice here, it lets a user specifically get the auto-sklearn wrapped version or the sklearn estimator which I suspect is of more relevance to people. No change needed, just commenting that I agree with it.

However I might still explicitly split out the tuple
```python
model_type, autosklearn_wrapped_model = model.steps[-1]
model_dict['sklearn_model'] = autosklearn_wrapped_model.choice.estimator
```

If the automatic checks give you a hard time with `model_type` not being used, you can do the following
```python
autosklearn_wrapped_model = model.steps[-1][1]
```